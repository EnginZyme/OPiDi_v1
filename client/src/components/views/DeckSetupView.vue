<template>
  <div>
    <div class="deck_setup_view">
      <div class="deck_setup_view__header">
        <div class="deck_setup_view__header__title">
          <span class="deck_setup_view__header__title__text">Deck Setup</span>
        </div>
      </div>
      <div class="deck_setup_view__body">
        <span>
          <label for="" class="deck_setup_view__body__device_type_label"
            >Device Type</label
          >
          <Dropdown
            :options="[{ name: 'Opentrons' }]"
            v-model="deviceType"
            optionLabel="name"
            placeholder="Select Device"
            class="deck_setup_view__body__dropdown"
          />
        </span>
        <div>
          <div class="deck_setup_view__body__deck_layout_card">
            <p>Deck Layout</p>
            <OpentronsSlotConfigForm />
          </div>
        </div>
        <div class="deck_setup_view__body__pipettes_card">
          <p>Pipettes</p>
          <div class="deck_setup_view__body__pipettes_card__grid">
            <div
              class="deck_setup_view__body__pipettes_card__grid__dropdown_group"
            >
              <label for="" class="deck_setup_view__body__pipettes_card__label"
                >Left Pipette</label
              >
              <Dropdown
                :options="pipettes"
                v-model="leftPipette"
                optionLabel="name"
                placeholder="Select Pipette"
                class="deck_setup_view__body__dropdown--long"
              />
            </div>
            <div
              class="deck_setup_view__body__pipettes_card__grid__dropdown_group"
            >
              <label for="" class="deck_setup_view__body__pipettes_card__label"
                >Right Pipette</label
              >
              <Dropdown
                :options="pipettes"
                v-model="rightPipette"
                optionLabel="name"
                placeholder="Select Pipette"
                class="deck_setup_view__body__dropdown--long"
              />
            </div>
            <div
              class="deck_setup_view__body__pipettes_card__grid__dropdown_group"
            >
              <label for="" class="deck_setup_view__body__pipettes_card__label"
                >Left Tipracks</label
              >
              <MultiSelect
                v-model="leftSelectedTipracks"
                :options="leftTipracks"
                optionLabel="id"
                placeholder="Select Tiprack(s)"
                class="deck_setup_view__body__dropdown--long"
              />
            </div>
            <div
              class="deck_setup_view__body__pipettes_card__grid__dropdown_group"
            >
              <label for="" class="deck_setup_view__body__pipettes_card__label"
                >Right Tipracks</label
              >
              <MultiSelect
                v-model="rightSelectedTipracks"
                :options="rightTipracks"
                optionLabel="id"
                placeholder="Select Tiprack(s)"
                class="deck_setup_view__body__dropdown--long"
              />
            </div>
          </div>
        </div>
        <br />
      </div>
    </div>
  </div>
</template>

<script>
import OpentronsSlotConfigForm from "../forms/OpentronsSlotConfigForm";
import { mapGetters } from "vuex";
import types from "../../types";

/**
 * A component that renders the deck setup view.
 */
export default {
  name: "DeckSetupView",
  components: {
    OpentronsSlotConfigForm,
  },
  data() {
    return {
      /**
       * Points the the robot device type object.
       */
      deviceType: {
        name: "Opentrons",
      },
    };
  },
  computed: {
    /**
     * A read/write object for the left pipette attribute of the protocol deck.
     */
    leftPipette: {
      get: function() {
        return this.pipettes.filter(
          (item) =>
            item.name === this.cachedProtocolDeck.left_pipette.pipette.name
        )[0];
      },
      set: function(newValue) {
        this.$store.dispatch(types.UPDATE_LEFT_PIPETTE, newValue);
      },
    },
    /**
     * A read/write object for the right pipette attribute of the protocol deck.
     */
    rightPipette: {
      get: function() {
        return this.pipettes.filter(
          (item) =>
            item.name === this.cachedProtocolDeck.right_pipette.pipette.name
        )[0];
      },
      set: function(newValue) {
        this.$store.dispatch(types.UPDATE_RIGHT_PIPETTE, newValue);
      },
    },
    /**
     * A read/write object for the left selected tipracks attribute of the protocol deck.
     */
    leftSelectedTipracks: {
      get: function() {
        return this.cachedProtocolDeck.left_pipette.tipracks.map((item) => ({
          id: item,
        }));
      },
      set: function(newValue) {
        this.$store.dispatch(
          types.UPDATE_LEFT_TIPRACKS,
          newValue.map((item) => item.id)
        );
      },
    },
    /**
     * A read/write object for the right selected tipracks attribute of the protocol deck.
     */
    rightSelectedTipracks: {
      get: function() {
        return this.cachedProtocolDeck.right_pipette.tipracks.map((id) => ({
          id,
        }));
      },
      set: function(newValue) {
        this.$store.dispatch(
          types.UPDATE_RIGHT_TIPRACKS,
          newValue.map((item) => item.id)
        );
      },
    },
    /**
     * Returns the tipracks available to be selected as left tipracks.
     */
    leftTipracks() {
      return this.chosenTipRacks
        .filter(
          (obj) =>
            !this.rightSelectedTipracks.filter((item) => item.id === obj.id)
              .length
        )
        .map((item) => ({
          id: item.id,
        }));
    },
    /**
     * Returns the tipracks available to be selected as right tipracks.
     */
    rightTipracks() {
      return this.chosenTipRacks
        .filter(
          (obj) =>
            !this.leftSelectedTipracks.filter((item) => item.id === obj.id)
              .length
        )
        .map((item) => ({
          id: item.id,
        }));
    },
    ...mapGetters(["pipettes", "chosenTipRacks", "cachedProtocolDeck"]),
  },
};
</script>

<style lang="scss" scoped>
.deck_setup_view__header__title__text {
  font-weight: 600;
  font-size: 18px;
  color: #040716;
  margin-left: 32px;
}
.deck_setup_view {
  min-height: 650px;
  min-width: 1000px;
  width: 95%;
  height: 90vh;
  background: #efefff;
  box-shadow: 6px 6px 9px 4px rgba(0, 0, 0, 0.25),
    -10px -10px 9px 4px rgba(255, 255, 255, 0.25);
  border-radius: 5px;
  align-self: center;
  justify-self: center;
  display: grid;
  grid-template-rows: auto 1fr;
  grid-row-gap: 2px;
}
.deck_setup_view__header {
  height: 48px;
  background: #efefff;
  box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.25);
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
  display: grid;
  align-content: center;
}
.deck_setup_view__body {
  display: grid;
  grid-template-rows: 40px auto auto auto;
  grid-row-gap: 30px;
  padding: 37px 40px 0px 40px;
  overflow: auto;
}
.deck_setup_view__body__device_type_label {
  margin-right: 15px;
  font-size: 14px;
}
.deck_setup_view__body__pipettes_card__label {
  font-size: 12px;
}
.deck_setup_view__body__dropdown {
  background: #efefff;
  width: 140px;
  font-size: 12px;
  box-shadow: 2px 2px 4px 1px rgba(0, 0, 0, 0.25),
    -2px -2px 4px 1px rgba(255, 255, 255, 0.25);
  border-radius: 6px;
}
.deck_setup_view__body__dropdown--long {
  background: #efefff;
  width: auto;
  font-size: 12px;
  box-shadow: 2px 2px 4px 1px rgba(0, 0, 0, 0.25),
    -2px -2px 4px 1px rgba(255, 255, 255, 0.25);
  border-radius: 6px;
}
.deck_setup_view__body__pipettes_card {
  width: 800px;
  height: 300px;
  display: grid;
  background: #efefff;
  box-shadow: -2px -2px 9px 2px rgba(255, 255, 255, 0.25),
    2px 2px 9px 2px rgba(0, 0, 0, 0.25);
  border-radius: 4px;
  padding: 1px 22px;
}
.deck_setup_view__body__pipettes_card__grid__dropdown_group {
  display: grid;
  grid-template-rows: 1.5em 32px;
}
.deck_setup_view__body__deck_layout_card {
  width: 100%;
  background: #efefff;
  box-shadow: -2px -2px 9px 2px rgba(255, 255, 255, 0.25),
    2px 2px 9px 2px rgba(0, 0, 0, 0.25);
  border-radius: 4px;
  padding: 1px 22px;
  margin-bottom: 10px;
}
.deck_setup_view__body__pipettes_card__grid {
  padding: 10px;
  padding-bottom: 8%;
  display: grid;
  grid-template-columns: auto auto;
  grid-column-gap: 50px;
  grid-row-gap: 1vh;
}
</style>
