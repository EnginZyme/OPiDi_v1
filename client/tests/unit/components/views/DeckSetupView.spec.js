import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import DeckSetupView from "@/components/views/DeckSetupView";

const localVue = createLocalVue();

localVue.use(Vuex);

const sampleCachedProtocolDeck = {
  left_pipette: {
    pipette: {
      name: "8 Channel 300 uL - Gen 2",
      id: "p300_multi_gen2",
      is_multichannel: true,
    },
    tipracks: [8],
  },
  right_pipette: {
    pipette: {
      name: "Single Channel 300 uL - Gen 1",
      id: "p300_single",
      is_multichannel: true,
    },
    tipracks: [7],
  },
  slots: {
    "1": {
      name: "wellplate_1",
      labware_name: "test 96 wellplate 2ml 3xT",
      labware_type: "shakerLabware",
      labware_id: "test_96_wellplate_2ml_3xT",
      nickname: "",
      display_name: "test 96 wellplate 2ml 3xT [1]",
      id: 1,
    },
    "2": {
      name: "tuberack_2",
      labware_name: "test tuberack 65 mL tubes",
      labware_type: "tubeRack",
      labware_id: "test_tuberack_65_mL_tubes",
      nickname: "",
      display_name: "test tuberack 65 mL tubes [2]",
      id: 2,
    },
    "3": {},
    "4": {},
    "5": {
      name: "reservoir_5",
      labware_name: "test slot 3 mL",
      labware_type: "reservoir",
      labware_id: "test_slot_3_mL",
      nickname: "",
      display_name: "test slot 3 mL [5]",
      id: 5,
    },
    "6": {},
    "7": {
      name: "tiprack_7",
      labware_name: "Opentrons 96 Tip Rack 0.3 mL",
      labware_type: "tipRack",
      labware_id: "opentrons_96_tiprack_300ul",
      nickname: "",
      display_name: "Opentrons 96 Tip Rack 0.3 mL [7]",
      id: 7,
    },
    "8": {
      name: "tiprack_8",
      labware_name: "Opentrons 96 Tip Rack 0.3 mL",
      labware_type: "tipRack",
      labware_id: "opentrons_96_tiprack_300ul",
      nickname: "",
      display_name: "Opentrons 96 Tip Rack 0.3 mL [8]",
      id: 8,
    },
    "9": {},
    "10": {},
    "11": {},
    "12": {},
  },
};
const sampleChosenTipRacks = [
  {
    name: "tiprack_7",
    labware_name: "Opentrons 96 Tip Rack 0.3 mL",
    labware_type: "tipRack",
    labware_id: "opentrons_96_tiprack_300ul",
    nickname: "",
    display_name: "Opentrons 96 Tip Rack 0.3 mL [7]",
    id: 7,
  },
  {
    name: "tiprack_8",
    labware_name: "Opentrons 96 Tip Rack 0.3 mL",
    labware_type: "tipRack",
    labware_id: "opentrons_96_tiprack_300ul",
    nickname: "",
    display_name: "Opentrons 96 Tip Rack 0.3 mL [8]",
    id: 8,
  },
];
const samplePipettes = [
  { name: "8 Channel 50 uL - Gen 1", id: "p50_multi", is_multichannel: true },
  { name: "8 Channel 300 uL - Gen 1", id: "p300_multi", is_multichannel: true },
  {
    name: "8 Channel 20 uL - Gen 2",
    id: "p20_multi_gen2",
    is_multichannel: true,
  },
  {
    name: "8 Channel 300 uL - Gen 2",
    id: "p300_multi_gen2",
    is_multichannel: true,
  },
  {
    name: "Single Channel 300 uL - Gen 1",
    id: "p300_single",
    is_multichannel: false,
  },
  {
    name: "Single Channel 50 uL - Gen 1",
    id: "p50_single",
    is_multichannel: false,
  },
];

describe("shallowMounted DeckSetupView", () => {
  let getters;
  let store;
  let wrapper;

  beforeEach(() => {
    getters = {
      cachedProtocolDeck: () => sampleCachedProtocolDeck,
      chosenTipRacks: () => sampleChosenTipRacks,
      pipettes: () => samplePipettes,
    };

    store = new Vuex.Store({
      getters,
    });

    wrapper = shallowMount(DeckSetupView, {
      store,
      localVue,
    });
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
