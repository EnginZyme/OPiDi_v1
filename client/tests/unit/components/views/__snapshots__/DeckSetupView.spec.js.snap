// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shallowMounted DeckSetupView renders correctly 1`] = `
<div>
  <div
    class="deck_setup_view"
  >
    <div
      class="deck_setup_view__header"
    >
      <div
        class="deck_setup_view__header__title"
      >
        <span
          class="deck_setup_view__header__title__text"
        >
          Deck Setup
        </span>
      </div>
    </div>
     
    <div
      class="deck_setup_view__body"
    >
      <span>
        <label
          class="deck_setup_view__body__device_type_label"
          for=""
        >
          Device Type
        </label>
         
        <dropdown
          class="deck_setup_view__body__dropdown"
          optionlabel="name"
          options="[object Object]"
          placeholder="Select Device"
        />
      </span>
       
      <div>
        <div
          class="deck_setup_view__body__deck_layout_card"
        >
          <p>
            Deck Layout
          </p>
           
          <opentronsslotconfigform-stub />
        </div>
      </div>
       
      <div
        class="deck_setup_view__body__pipettes_card"
      >
        <p>
          Pipettes
        </p>
         
        <div
          class="deck_setup_view__body__pipettes_card__grid"
        >
          <div
            class="deck_setup_view__body__pipettes_card__grid__dropdown_group"
          >
            <label
              class="deck_setup_view__body__pipettes_card__label"
              for=""
            >
              Left Pipette
            </label>
             
            <dropdown
              class="deck_setup_view__body__dropdown--long"
              optionlabel="name"
              options="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
              placeholder="Select Pipette"
            />
          </div>
           
          <div
            class="deck_setup_view__body__pipettes_card__grid__dropdown_group"
          >
            <label
              class="deck_setup_view__body__pipettes_card__label"
              for=""
            >
              Right Pipette
            </label>
             
            <dropdown
              class="deck_setup_view__body__dropdown--long"
              optionlabel="name"
              options="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
              placeholder="Select Pipette"
            />
          </div>
           
          <div
            class="deck_setup_view__body__pipettes_card__grid__dropdown_group"
          >
            <label
              class="deck_setup_view__body__pipettes_card__label"
              for=""
            >
              Left Tipracks
            </label>
             
            <multiselect
              class="deck_setup_view__body__dropdown--long"
              optionlabel="id"
              options="[object Object]"
              placeholder="Select Tiprack(s)"
            />
          </div>
           
          <div
            class="deck_setup_view__body__pipettes_card__grid__dropdown_group"
          >
            <label
              class="deck_setup_view__body__pipettes_card__label"
              for=""
            >
              Right Tipracks
            </label>
             
            <multiselect
              class="deck_setup_view__body__dropdown--long"
              optionlabel="id"
              options="[object Object]"
              placeholder="Select Tiprack(s)"
            />
          </div>
        </div>
      </div>
       
      <br />
    </div>
  </div>
</div>
`;
